### PR TITLE
Remove legacyFactor and adopt blackHole where appropriate in Diffing and Myers benchmarks

### PR DESCRIPTION
**Pull request by numist**
Per post-merge feedback in https://github.com/apple/swift/pull/25808

### CODE REVIEW

**Comment by numist** [comment]
@swift-ci please benchmark

**Comment by numist** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
The benchmark validation on CI runs only for the newly added tests, so we don’t see those results here now. Could you please also post your local result from running `Benchmark_Driver check -f Diff -f Myers --format markdown`? Otherwise LGTM.

**Review by palimondo — CHANGES_REQUESTED** [review]
Could you please refactor to use a shared run function, parametrized with the workload String? You should probably also touch the lazily initialized global constants in setUpFunction using the blackHole, too. See AngryPhonebook benchmarks for an example of this.

**Comment by numist** [comment]
✅  | Benchmark Check Report
---|---
⚠️⏱ | `Myers` execution took at least 8040 μs. <br><sub> Decrease the workload of `Myers` by a factor of 16 (10), to be less than 1000 μs.
⚠️Ⓜ️ | `Myers` has very wide range of memory used between independent, repeated measurements. <br><sub> `Myers` mem_pages [i1, i2]: min=[2140, 2672] 𝚫=532 R=[1074, 569]

After touching the globals and adjusting the inputs of `Myers` to match `Diffing.Similar` in 16087175b7:

✅  | Benchmark Check Report
---|---
⚠️Ⓜ️ | `Myers` has very wide range of memory used between independent, repeated measurements. <br><sub> `Myers` mem_pages [i1, i2]: min=[35, 43] 𝚫=8 R=[20, 41]

Not sure why `Myers` would exhibit this problem and not `Diffing.Similar` as the two are practically identical (until I land further perf improvements to `Collection.difference(from:)`, anyway).

As an aside, could you help me understand why everything gets `blackHole`d? I understand why you'd want to mask out setup and the like, but AngryPhoneBook even blackholes the work that is supposed to be benchmarked so it seems like I don't have the right idea around what it's for.

**Comment by palimondo** [comment]
As I understand it, the`blackHole` is used as barrier to prevent sufficiently smart future compiler from eliminating calls with unused results as dead code. But thinking about this particular case… the `let _ = ` should guarantee the same thing, right? 🤔 @lorentey, @eeckstein?

We started to use it in `setUpFunction`s to pre-initialize lazy globals that take significant time, so that their initialization does not spoil the first sample.

**Comment by numist** [comment]
Sounds reasonable. Regardless, there's value in conforming to the existing patterns.

**Comment by numist** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
> `Myers` mem_pages [i1, i2]: min=[35, 43] 𝚫=8 R=[20, 41]

This seems fine to me. The memory heuristic is quite sensitive.

**Review by palimondo — CHANGES_REQUESTED** [review]
Scott, I apologize for the being skittish in my review… I’m on the road and doing all  this on a phone.

Could you please also change the naming of these benchmarks to follow the [Naming Convention](http://bit.ly/BenchmarkNaming)? I think the benchmark family name here is clearly `Diff`, so putting a period after that would we a good first step. If I understand this correctly, Myers is an alternative diffing algorithm. In your last commit, you’ve changed the workloads in `Myers` to match those in `DiffSimilar`. Therefore it would make sense to try and make the results directly comparable by also applying the concept of variants from the benchmark naming convention.  Could you also please:

* change the variable names of workloads  in Myers.swift to match those from Diffing.swift, so that it is clearer it is the same workload,
* rename the file Myers.swift to DiffingMyers.swift or DiffMyers.swift,
* rename `Myers` benchmark to `Diff.Similar.Myers`, to make it clear it is a variant of the same workload, using a different algorithm.

**Comment by palimondo** [comment]
Tiny nitpick: I’d declare the reversed workloads like this:
````swift
let alphabetsReversed = Array(alphabets.reversed())
let loremReverse = Array(loremIpsum.reversed())
````
…and probably call the latter <code>loremReverse<strong>d</strong></code>… 😜

**Comment by numist** [comment]
@swift-ci Please test

**Comment by palimondo** [comment]
The changes in `DiffingMeyers` and `Diffing` look great, but something went very wrong during the merge. It now has 148 changed files instead of 3.

**Comment by numist** [comment]
Something was extremely broken with my fork. Should be fixed now.

**Comment by palimondo** [comment]
@swift-ci please benchmark

File: benchmark/single-source/DiffingMyers.swift [hunk 1]
```diff
@@ -12,18 +12,25 @@
 
 import TestsUtils
 
-public let Myers = [
-  BenchmarkInfo(name: "Myers", runFunction: run_Myers, tags: [.algorithm]),
-]
+// The DiffingMyers test benchmarks Swift's performance running the algorithm
```
Comment (palimondo) [inline benchmark/single-source/DiffingMyers.swift]: Thanks for adding this explanation!

**Review by palimondo — APPROVED** [review]
LGTM. Thank you for your patience Scott!

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by numist** [comment]
How did @swift-ci come up with zero run time for the diffing benchmarks after the local benchmark check report indicated everything was fine?

**Comment by numist** [comment]
Ohhh, I bet it was not actually appropriate to use `blackHole` around the work function.

**Comment by lorentey** [comment]
At worst, `blackHole` should slow things down, not speed things up. The use of `blackHole` in the `run_` functions seems entirely appropriate to me. 

(It looks somewhat out of place in the setup closures, but it won't cause harm there, either.)

**Comment by lorentey** [comment]
What's likely happening is that the optimizer is clever enough to recognize that the same input is repeatedly passed to the same pure function, and it moves the function call outside the loop. If so, the `identity` helper should prevent that:

```
blackHole(diff(identity(arg1), identity(arg2))
```

**Comment by lorentey** [comment]
This highlights an interesting point -- since the Myers algorithm is implemented internal to the benchmark, it has an unfair advantage over the stdlib: it can be specialized to the concrete element type, which can considerably speed things up.

**Comment by palimondo** [comment]
I’m pretty sure the zero runtimes are caused by CI machines running on stable macOS 10.14 and not the 10.15 Beta, so it is the expected result of the availability checks we’ve been discussing in  #25808.

**Comment by numist** [comment]
Oh I see, the `blackHole` function prevents code elimination by the compiler.

> This highlights an interesting point -- since the Myers algorithm is implemented internal to the benchmark, it has an unfair advantage over the stdlib: it can be specialized to the concrete element type, which can considerably speed things up.

I think this is ok since the Myers benchmark measures Swift vs the Diffing benchmark, which measures `Collection.difference()`

> I’m pretty sure the zero runtimes are caused by CI machines running on stable macOS 10.14 and not the 10.15 Beta, so it is the expected result of the availability checks we’ve been discussing in #25808.

That makes sense, especially since that I got no such warnings from `Benchmark_Driver check` on a 10.15 system. We could put the `BenchmarkInfo` instances inside an availability check to omit them from the results, or just revisit this once 10.15 has GMed if the zeroes aren't hurting anything.

**Comment by palimondo** [comment]
The zeros are not hurting anything directly, it’s just that these benchmarks also provide no value here until the machines upgraded to 10.15. So I still question the way we do the availability checks in these benchmarks. 

What @lorentey explained in #25808 makes sense for app developers using the new diffing API, but I’m not convinced it applies equally to these benchmarks. I think the Swift Benchmark Suite’s use and usefulness outside of guarding compiler and stdlib development is rather theoretical and should not be used as argument to hamper this primary use case.

Can we please re-examine this issue? CC @eeckstein @jrose-apple

**Pull request merged**